subgraphs/compound-v3/protocols/compound-v3/src/mapping.ts

```diff
@@ -0,0 +1,1210 @@
+  updateMarketData(market);
+  updateRevenue(market, event.address);
+  updateRewards(market, event.address, event);
+
+  let newBalance = getUserBalance(
+    cometContract,
+    event.params.from,
+    null,
+    PositionSide.COLLATERAL
+  );
+  market.createWithdraw(
+    tryBaseToken.value,
+    event.params.from,
+    amount,
+    amount
+      .toBigDecimal()
+      .div(exponentToBigDecimal(token.decimals))
+      .times(token.lastPriceUSD!),
+    newBalance,
+    InterestRateType.VARIABLE
+  );
+
+  amount = ethereum.decode("uint256", supplyLog.data)!.toBigInt();
+  const toAddress = ethereum
+    .decode("address", supplyLog.topics.at(2))!
+    .toAddress();
+  newBalance = getUserBalance(
+    cometContract,
+    toAddress,
+    null,
+    PositionSide.COLLATERAL
+  );
+  market.createDeposit(
+    tryBaseToken.value,
+    toAddress,
+    amount,
+    amount
+      .toBigDecimal()
+      .div(exponentToBigDecimal(token.decimals))
+      .times(token.lastPriceUSD!),
+    newBalance,
+    InterestRateType.VARIABLE
+  );
+}
+
+//
+//
+// Transfer user collateral to another account
+export function handleTransferCollateral(event: TransferCollateral): void {
+  const marketID = event.address.concat(event.params.asset);
+  const market = new DataManager(
+    marketID,
+    event.params.asset,
+    event,
+    getProtocolData()
+  );
+  const sender = event.params.from;
+  const receiver = event.params.to;
+  const asset = event.params.asset;
+  const amount = event.params.amount;
+  const token = market.getInputToken();
+  updateMarketData(market);
+  // no revenue accrued during this event
+  updateRewards(market, event.address, event);
+
+  const cometContract = Comet.bind(event.address);
+  const senderBalance = getUserBalance(cometContract, sender, asset);
+  const receiverBalance = getUserBalance(cometContract, receiver, asset);
+  market.createTransfer(
+    asset,
+    sender,
+    receiver,
+    amount,
+    bigIntToBigDecimal(amount, token.decimals).times(token.lastPriceUSD!),
+    senderBalance,
+    receiverBalance
+  );
+}
+
+//
+//
+// Sell liquidated collateral at a discount (of liquidation penalty)
+export function handleAbsorbCollateral(event: AbsorbCollateral): void {
+  const marketID = event.address.concat(event.params.asset);
+  const market = new DataManager(
+    marketID,
+    event.params.asset,
+    event,
+    getProtocolData()
+  );
+  const marketEntity = market.getMarket();
+  const cometContract = Comet.bind(event.address);
+  const liquidator = event.params.absorber;
+  const borrower = event.params.borrower;
+  const baseAsset = cometContract.baseToken();
+  const amount = event.params.collateralAbsorbed;
+  const amountUSD = bigIntToBigDecimal(amount, COMPOUND_DECIMALS);
+  const liquidationPenalty =
+    marketEntity.liquidationPenalty.div(BIGDECIMAL_HUNDRED);
+  const profitUSD = amountUSD.times(liquidationPenalty);
+  updateMarketData(market);
+  updateRevenue(market, event.address);
```

Comment:
Not familar with Compound v3, so forgive my stupid questions here: Is `event.params.collateralAbsorbed` in USD with COMPOUND_DECIMALS? Does liquidation profit (liquidationPenalty) only go to the liquidator?

Reply:
Good catch, there is actually a param called `usdValue` that is 8 decimals offset